### PR TITLE
Improve consistency in UNICODEMAP code and docs, update docs/understanding_qmk

### PR DESCRIPTION
**Pull request by vomindoraan**
Rename some constants and functions pertaining to UNICODEMAP so they're consistent with the rest of the system. Also update the documentation accordingly. See below for details.

These changes should be non-breaking. I've taken care to update all usages of the renamed symbols.

## Description
- Remove unused `UNICODE(n)` macro (alias for `UC(n)`, but no one was using it).
- Rename `QK_UNICODE_MAP` constant to `QK_UNICODEMAP` (only used internally).
- Rename `unicode_map_input_error` function to `unicodemap_input_error` (only used internally).
- Rename `process_unicode_map` function to `process_unicodemap` (only used internally).

  **Reasoning:** `UNICODEMAP` is the name of the [keymap definition method](https://github.com/qmk/qmk_firmware/blob/1d200b607b4722e7f70af5668a8849a453fdb1a2/docs/feature_unicode.md#unicode-support) (hence `UNICODEMAP_ENABLE`, `process_unicodemap` etc.), whereas `unicode_map` refers to the mapping table itself, which is used by the method.

- Refactor `process_unicodemap` code a bit to make it more readable.
- Update links in [docs/understanding_qmk](https://github.com/qmk/qmk_firmware/blob/1d200b607b4722e7f70af5668a8849a453fdb1a2/docs/understanding_qmk.md#process-record) and rearrange the list of process record functions so it matches the actual order in which they're called now (the order was slightly altered in #4325).
- Various Unicode doc improvements.

Tested by compiling several keymaps that use UNICODEMAP.

P.S. When I say it's only used internally, I mean it's only used by QMK's internals and not directly in any userspaces/keymaps. That's why I think it's okay to rename these symbols. No “user facing” symbols were altered in the making of this PR 😛 

## Types of changes
- [ ] Core
- [ ] Bugfix
- [ ] New Feature
- [x] Enhancement/Optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/Layout/Userspace (addition or update)
- [x] Documentation

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/quantum_keycodes.h [hunk 1]
```diff
@@ -679,15 +679,12 @@ enum quantum_keycodes {
...
-    #define UNICODE(n) (QK_UNICODE | (n))
-    #define UC(n) UNICODE(n)
+  // Allows Unicode input up to 0x7FFF
+  #define UC(c) (QK_UNICODE | (c))
 #endif
-
 #ifdef UNICODEMAP_ENABLE
-    #define X(n) (QK_UNICODE_MAP | (n))
+  // Allows Unicode input up to 0x10FFFF, requires unicode_map
+  #define X(i) (QK_UNICODEMAP | (i))
```
Comment (drashna) [inline quantum/quantum_keycodes.h]: I forgot to add this myself, but I'd like to see an "else" for both of these (`UC` and `X`) that define them as empty functions, so they don't error out when compiling when the respective feature is disabled.

Comment (vomindoraan) [reply quantum/quantum_keycodes.h]: They could be defined as `KC_NONE` in that case. But what's the use case? It seems to me like you'd want it to error out as fast as possible if you put `UC` or `X` in your keymap but forget to enable the corresponding feature.

Comment (drashna) [reply quantum/quantum_keycodes.h]: I'm not sure that failing is the best option here, to be honest. 

Two reasons really come up for this: 
* Community Layouts 
* EEPROM storage of keymaps

Also, testing comes to mind, but that's an edge case. 

The alternative is that if I want to do this locally, I would have to "ifdef" the keycodes ANYWAYS.

Comment (drashna) [reply quantum/quantum_keycodes.h]: But for now, I'll just merge this.

**Comment by drashna** [comment]
Thanks!

File: quantum/quantum_keycodes.h (hunk 1 above)
Comment (vomindoraan) [reply quantum/quantum_keycodes.h]: Could you elaborate on the community layouts bit? I think I know what you mean, but it'd be best if you could clarify just in case.

As far as keymap storage goes, `UC` and `X` themselves aren't part of the keycodes enum (they're just macros), so leaving them undefined doesn't affect other keycodes. What does get left out is `QK_UNICODE` and `QK_UNICODEMAP`, but if I'm not mistaken, those _should_ be left out for the sake of feature flagging.
